### PR TITLE
feat(EMS-895): no PDF - CRON - Abandoned application access

### PR DESCRIPTION
**Pull request by Zainzzkk**
## Introduction :pencil2:
This PR prevents access to an abandoned application, including in the dashboard.

## Resolution :heavy_check_mark:
* Updated UI graphql query to filter out Abandoned applications
* Updated API to create and update an abandoned application
* Split up createAnApplication to use a helper to create the application
* Added createAnAbandonedApplication to allowed dev resolvers
* Added new e2e function and api call for creating abandoned application
* Added test for access to an abandoned application

## Miscellaneous :heavy_plus_sign:
* Updated ui eslintrc.js
* Lint fixes

### CODE REVIEW

File: src/api/custom-resolvers/mutations/create-an-abandoned-application/index.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+  // set status to abandoned
+  abandonedApplicationVariables.status = STATUS.ABANDONED;
+
+  try{
+    // creates and returns application
+    const createdApplication = await createAnApplicationHelper(root, abandonedApplicationVariables, context);
+
+    if (createdApplication) {
+      // TODO: remove once schema hooks for status is removed
```
Comment (ttbarnes) [inline src/api/custom-resolvers/mutations/create-an-abandoned-application/index.ts:30]: Can we add `EMS-3387` ticket number here

File: e2e-tests/commands/insurance/create-an-abandoned-application.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+       * pass sectionReview for eligibility to API
+       * has to be set to true for eligibility
+       */
+      const sectionReview = {
+        eligibility: true,
+      };
+
+      return api.createAnAbandonedApplication(accountId, mockEligibilityAnswers, company, sectionReview).then((application) => application);
+    } catch (err) {
+      console.error('Creating an application', err);
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/create-an-abandoned-application.js:34]: Abandoned* application?

File: e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination/dashboard-pagination-15-applications.spec.js [hunk 3]
```diff
@@ -11,13 +11,19 @@ const dashboardUrl = `${baseUrl}${DASHBOARD}`;
...
   let applications;
+  let abandonedApplication;
 
   before(() => {
     cy.completeSignInAndGoToDashboard().then(({ accountId }) => {
       cy.createApplications(accountId, totalApplications).then((createdApplications) => {
         applications = createdApplications;
       });
 
+      // creates an abandoned application
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination/dashboard-pagination-15-applications.spec.js:22]: I don't think we need this comment, self explanatory

Could we remove the abandoned application from this test and create a new dashboard/pagination test where there are 2 applications, one abandoned, one regular "in progress" application, and then assert the length and status via:

```js
cy.assertLength(table.body.rows(), 1);
dashboardPage.table.body.row(referenceNumber)
```

Should probably add a test for 17 applications, with one of them being abandoned. So the pagination is rendered, but no abandoned applications are present

Then the existing E2E test can remain as a happy path

File: e2e-tests/insurance/cypress/e2e/journeys/no-access-to-application/no-access-to-application-abandoned.spec.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+const {
+  ROOT,
+  ALL_SECTIONS,
+  NO_ACCESS_TO_APPLICATION,
+} = INSURANCE_ROUTES;
+
+const baseUrl = Cypress.config('baseUrl');
+
+context('Insurance - no access to application page - Abandoned application', () => {
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/no-access-to-application/no-access-to-application-abandoned.spec.js:11]: Do we have user story type copy to add add the end here?

File: e2e-tests/insurance/cypress/e2e/journeys/no-access-to-application/no-access-to-application-abandoned.spec.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  let refNumber;
+  let applicationUrl;
+
+  before(() => {
+    cy.saveSession();
+
+    cy.completeSignInAndGoToDashboard().then(({ accountId }) => {
+      cy.createAnAbandonedApplication(accountId).then((createdApplication) => {
+        applicationUrl = `${baseUrl}${ROOT}/${createdApplication.referenceNumber}${ALL_SECTIONS}`;
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/no-access-to-application/no-access-to-application-abandoned.spec.js:20]: `allSectionsUrl`?

File: e2e-tests/insurance/cypress/e2e/journeys/no-access-to-application/no-access-to-application-abandoned.spec.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        refNumber = createdApplication.referenceNumber;
+      });
+    });
+  });
+
+  after(() => {
+    cy.deleteApplication(refNumber);
+  });
+
+  describe('when trying to access an Abandoned application', () => {
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/no-access-to-application/no-access-to-application-abandoned.spec.js:30]: `ABANDONED` constant?

File: src/api/custom-resolvers/mutations/create-an-abandoned-application/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  describe('when creation is not successful', () => {
+    test('it should throw an error', async () => {
+      try {
+        // pass empty context object to force an error
+        await createAnAbandonedApplication({}, variables, {});
+      } catch (err) {
+        const errorString = String(err);
+
+        expect(errorString.includes('Creating an abandoned application')).toEqual(true);
+      }
```
Comment (ttbarnes) [inline src/api/custom-resolvers/mutations/create-an-abandoned-application/index.test.ts:76]: `rejects.toThrow(...)` ?

File: src/api/custom-resolvers/mutations/create-an-abandoned-application/index.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ * 1) Set status to Abandoned
+ * 2) Create a new application with createAnApplicationHelper.
+ * 3) Updates application with abandoned status.
+ * 4) Returns application and success flag
+ * @param {Object} GraphQL root variables
+ * @param {CreateAnApplicationVariables} GraphQL variables for the CreateAnAbandonedApplication mutation
+ * @param {Context} KeystoneJS context API
+ * @returns {Promise<Object>} Object with success flag and application
+ */
+const createAnAbandonedApplication = async (root: any, variables: CreateAnApplicationVariables, context: Context) => {
```
Comment (ttbarnes) [inline src/api/custom-resolvers/mutations/create-an-abandoned-application/index.ts:19]: Can we add a `console.info` here?

File: src/api/custom-resolvers/mutations/create-an-application/index.ts [hunk 9]
```diff
@@ -1,143 +1,42 @@
...
  * @param {Object} GraphQL root variables
- * @param {Object} GraphQL variables for the CreateAnApplication mutation
- * @param {Object} KeystoneJS context API
- * @returns {Promise<Object>} Object with success flag
+ * @param {CreateAnApplicationVariables} GraphQL variables for the createAnApplication mutation
+ * @param {Context} KeystoneJS context API
+ * @returns {Promise<Object>} Object with success flag and application
  */
 const createAnApplication = async (root: any, variables: CreateAnApplicationVariables, context: Context) => {
   console.info('Creating application for ', variables.accountId);
```
Comment (ttbarnes) [inline src/api/custom-resolvers/mutations/create-an-application/index.ts:20]: RE above, like this

File: src/api/custom-resolvers/mutations/create-an-abandoned-application/index.test.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    expect(result.success).toEqual(true);
+  });
+
+  test(`it should return status as ${STATUS.ABANDONED}`, async () => {
+    result = await createAnAbandonedApplication({}, variables, context);
+
+    expect(result.status).toEqual(STATUS.ABANDONED);
+  });
+
+  test(`it should return submissionType as ${SUBMISSION_TYPE.MIA}`, async () => {
```
Comment (ttbarnes) [inline src/api/custom-resolvers/mutations/create-an-abandoned-application/index.test.ts:51]: I don't believe we need this test since the helper handles this,

Same for `create-an-application` GQL mutation

File: src/api/custom-resolvers/mutations/create-an-application/index.test.ts [hunk 11]
```diff
@@ -39,46 +42,16 @@ describe('custom-resolvers/create-an-application', () => {
...
-    const createdApplication = await applications.get({ context, applicationId: result.id });
-
-    const expected = createdApplication.referenceNumber;
-
-    expect(result.referenceNumber).toEqual(expected);
+    expect(result.status).toEqual(STATUS.IN_PROGRESS);
   });
 
-  test('it should return application relationships', async () => {
+  test(`it should return submissionType as ${SUBMISSION_TYPE.MIA}`, async () => {
```
Comment (ttbarnes) [inline src/api/custom-resolvers/mutations/create-an-application/index.test.ts:51]: Same as above

File: src/api/custom-schema/type-defs.ts [hunk 12]
```diff
@@ -205,6 +205,13 @@ const typeDefs = `
     referenceNumber: Int
   }
 
+  type CreateAnAbandonedApplicationResponse {
+    success: Boolean!
+    id: String
+    referenceNumber: Int
+    status: String
```
Comment (ttbarnes) [inline src/api/custom-schema/type-defs.ts:212]: Is there a reason for having `status` here? As we don't have this for `CreateAnApplicationResponse`

File: src/ui/server/graphql/queries/applications.ts [hunk 13]
```diff
@@ -2,7 +2,12 @@ import gql from 'graphql-tag';
 
 const applicationsQuery = gql`
   query applications($accountId: ID!, $take: Int!, $skip: Int!) {
-    applications(where: { owner: { id: { equals: $accountId } } }, orderBy: { updatedAt: desc }, take: $take, skip: $skip) {
+    applications(
+      where: { AND: [{ owner: { id: { equals: $accountId } } }, { status: { not: { equals: "Abandoned" } } }] }
+      orderBy: { updatedAt: desc }
+      take: $take
+      skip: $skip
+    ) {
```
Comment (ttbarnes) [inline src/ui/server/graphql/queries/applications.ts:10]: @abhi-markan FYI these is the GQL syntax/variables RE discussions the other day 😎

File: src/ui/server/middleware/insurance/application-status/index.ts [hunk 14]
```diff
@@ -37,6 +44,14 @@ export const applicationStatusMiddleware = async (req: Request, res: Response, n
         return res.redirect(NO_ACCESS_APPLICATION_SUBMITTED);
       }
 
+      /**
```
Comment (ttbarnes) [inline src/ui/server/middleware/insurance/application-status/index.ts:47]: Nice, can we update the documentation at the top on line 19?

File: src/api/custom-schema/type-defs.ts (hunk 12 above)
Comment (Zainzzkk) [reply src/api/custom-schema/type-defs.ts:212]: had it there for testing.  Removed

File: e2e-tests/insurance/cypress/e2e/journeys/no-access-to-application/no-access-to-application-abandoned.spec.js (hunk 4 above)
Comment (Zainzzkk) [reply e2e-tests/insurance/cypress/e2e/journeys/no-access-to-application/no-access-to-application-abandoned.spec.js:11]: done

File: e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-abandoned-applications.spec.js [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+  });
+
+  after(() => {
+    cy.deleteApplications(applications);
+    cy.deleteApplication(abandonedApplication.referenceNumber);
+  });
+
+  it(`should render 1 ${IN_PROGRESS} application on the dashboard`, () => {
+    cy.assertLength(table.body.rows(), totalApplications);
+    dashboardPage.table.body.row(applications[0].referenceNumber).status();
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-abandoned-applications.spec.js:46]: Need to use `cy.checkText` with an actual status here 😛 

Can we also rename `totalApplications` to `totalExpectedApplications`?

File: e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination/dashboard-pagination-16-applications-and-1-abandoned.spec.js [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+const totalApplications = MAX_APPLICATIONS_PER_PAGE + 1;
+const totalPages = 2;
+
+const dashboardUrl = `${baseUrl}${DASHBOARD}`;
+
+const { table } = dashboardPage;
+
+const { ABANDONED } = APPLICATION.STATUS;
+
+context(`Insurance - Dashboard - pagination - ${totalApplications} applications and 1 abandoned application`, () => {
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination/dashboard-pagination-16-applications-and-1-abandoned.spec.js:19]: could we use `ABANDONED` constant here?

File: e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination/dashboard-pagination-16-applications-and-1-abandoned.spec.js [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+    it('should have the correct pagination state', () => {
+      cy.assertPaginationState({
+        totalPages,
+        index: 0,
+        expectedPageNumber: 1,
+        previousLinkShouldExist: false,
+        expectedUrl: `${baseUrl}${DASHBOARD}`,
+      });
+    });
+  });
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination/dashboard-pagination-16-applications-and-1-abandoned.spec.js:67]: Could we add a test to make sure none of the application's on the first page have an `ABANDONED` status? Something like this:

```js
dashboardPage.body.rows().forEach((row) => {
  cy.checkText(row.status(), IN_PROGRESS);
});
```

File: e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination/dashboard-pagination-16-applications-and-1-abandoned.spec.js [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+        totalPages,
+        index: 2,
+        expectedPageNumber: 2,
+        nextLinkShouldExist: false,
+      });
+    });
+
+    it(`should have 1 application on the dashboard (and not the ${ABANDONED} application)`, () => {
+      cy.assertLength(table.body.rows(), 1);
+    });
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination/dashboard-pagination-16-applications-and-1-abandoned.spec.js:93]: This isn't actually testing the status, can we update to use the `rows().forEach((row) => { ...` above?

Could move this into a cypress command maybe `assertAllDashboardRowsHaveStatusInProgress`

File: e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-abandoned-applications.spec.js [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+  });
+
+  after(() => {
+    cy.deleteApplications(applications);
+    cy.deleteApplication(abandonedApplication.referenceNumber);
+  });
+
+  it(`should render 1 ${IN_PROGRESS} application on the dashboard`, () => {
+    cy.assertLength(table.body.rows(), totalExpectedApplications);
+    cy.checkText(dashboardPage.table.body.row(applications[0].referenceNumber).status(), IN_PROGRESS);
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-abandoned-applications.spec.js:46]: Not a blocker, could use `cy.createAnApplication` instead of `cy.createApplications` and then use `referenceNumber` like other non-pagination tests, since there's only 1 application that's being tested here

**Review by ttbarnes — APPROVED** [review]
Great work
